src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -231,28 +252,67 @@ case class WorkflowActor(workflow: WorkflowDescriptor,
+       * Each method is tried individually and the first to return a Success value takes precedence.
        */
 
-      val namespaces = workflow.namespace.namespaces filter {_.importedAs.contains(identifierString)}
-      namespaces.headOption.getOrElse {
-        val matchedCall = parentWorkflow.calls.find {_.name == identifierString}.getOrElse {
-          throw new WdlExpressionException(s"Expecting to find a call with name '$identifierString'")
+      def lookupNamespace(name: String): Try[WdlNamespace] = {
+        workflow.namespace.namespaces.find {_.importedAs.contains(name)} match {
+          case Some(x) => Success(x)
+          case _ => Failure(new WdlExpressionException(s"Could not resolve $identifierString as a namespace"))
         }
-        val futureValue = fetchCallOutputEntries(matchedCall).map { entries =>
-          val callOutputs = entries.map { entry =>
-            val value = entry.wdlValue match {
-              case Some(v) => v
-              case _ => throw new WdlExpressionException(s"Could not evaluate call '${matchedCall.name}', because '${entry.key.name}' is undefined")
+      }
+
+      def lookupCall(name: String): Try[WdlObject] = {
+        parentWorkflow.calls.find {_.name == identifierString} match {
+          case Some(matchedCall) =>
+            val futureValue = fetchCallOutputEntries(matchedCall).map {callOutputEntries =>
+              val callOutputsAsMap = callOutputEntries.map { entry =>
+                entry.key.name -> entry.wdlValue
+              }.toMap
+              callOutputsAsMap.find {case (k,v) => v.isEmpty} match {
+                case Some(noneValue) => Failure(new WdlExpressionException(s"Could not evaluate call ${matchedCall.name} because some if its inputs are not defined (i.e. $noneValue._1"))
+                case None => Success(WdlObject(callOutputsAsMap.map {case (k,v) => k -> v.get}))
+              }
             }
-            entry.key.name -> value
-          }
-          WdlObject(callOutputs.toMap)
+            Await.result(futureValue, DatabaseTimeout)
+          case None => Failure(new WdlExpressionException(s"Could not find a call with name '$identifierString'"))
         }
-        Await.result(futureValue, DatabaseTimeout)
       }
+
+      def lookupDeclaration(name: String): Try[WdlValue] = {
+        parentWorkflow.declarations.find {_.name == identifierString} match {
+          case Some(declaration) =>
+            val futureValue = fetchFullyQualifiedName(declaration.fullyQualifiedName).map {
+              case t: Traversable[SymbolStoreEntry] if t.isEmpty =>
+                Failure(new WdlExpressionException(s"Could not find a declaration with name '$identifierString'"))
+              case t: Traversable[SymbolStoreEntry] if t.size > 1 =>
+                Failure(new WdlExpressionException(s"Expected only one declaration for '$identifierString', got ${t.size}"))
+              case t: Traversable[SymbolStoreEntry] => t.head.wdlValue match {
```

Comment:
What I mean is to have a function which takes `declaration` (and I think `identifierString`) which gets called by the `map` and then `getOrElse` returns your `Failure` clause.